### DISCUSSION TITLE
Whitebox Manipulator Mode Suggestion And Improvements

### DISCUSSION THREAD

**Discussion by pollend** [o3de/o3de#12822] [Ideas]
recently merged into whitebox to improve functionality: https://github.com/o3de/o3de/pull/9925
This feature is merged in but currently disabled but can be enabled under `<project>/user/Registry/editorprefences.setreg`
```
"O3DE": {
   "Preferences": {
       "WhiteBox": {
               "TransformFeature": true
```

Any comments to improve the functionality of whitebox can be discussed here.

**Comment by hultonha** [comment]
Feedback from @LB-ArturZieba 

> I like it, it makes the White Box editing more flexible than the Sketch mode, however I have two suggestions if you would expand this past the experimental phase:
> 1. A way of adjusting the Translation/Rotation/Scale of Vertices/Edges/Scale exactly, similar to how it works on the Transform component. Currently if I want to for example rotate an edge exactly 45 degrees along some axis I need to use Angle Snapping with that value to achieve that.
> 1. A way of merging multiple vertices that are on the same coordinates. This is something that could speed up regular workflows and can be found for example in Blender as a "Merge by Distance" option.

> I have also noticed an unintuitive behavior while investigating the https://github.com/o3de/o3de/issues/12816. The scaling of faces along the perpendicular is not possible. If such plane is rotated so that it no longer is perpendicular to the said axis, it can be scaled.
> While it makes sense for the 0-width plane to not be scalable along perpendicular axis, the fact that they become scalable after a certain rotation (as if gaining that width) causes confusion.
> Should this be reported as an additional issue, or is it a feature still expanded on, for example adding a scaling option based on face origin instead of the global transform?

Feedback from @LB-MateuszSzweda 

> I also like the direction. I was able to shape White Boxes more precisely and definitely into more sophisticated shapes than before.
Before the change when working with White Box I found myself working with multiple entities and only by morphing and overlapping them I was able to achieve desired results.
> Now it was possible to achieve these shapes with only one Entity.
> Apart from things mentioned from Artur I think the last thing White box need to achieve a fully customizable, malleable experience is a way to add new vertices to edges, similarly as in the Spline component.

> When it comes to the scope of testing.
> - I have verified that all of white boxes default shapes and custom meshes could be modified with manipulator mode.
> - Multiple White box entities could be modified at once (however modifying more than 3 with manipulator mode quickly clutters the Viewport edge and makes switching between modes much harder)
> - Manipulator modified White boxes could be exported as obj. and saved as white box meshes which could be assigned.
> - Manipulator modified White boxes could be saved as prefabs and each instantiated prefab reacted to further changes by manipulator mode.
> - Editor remained stable when manipulating very large White boxes with large amount of edges and vertices.
> - Manipulator modified white boxes could have a White Box collider assigned. It was recognized by PhysX_debug and correctly interacted with PhysX entities.

**Comment by hultonha** [comment]
Also one other point of discussion that came up is we need a new icon for this mode. @amzn-leenguy would you please be able to help with this?

@pollend Could you please attach a screen shot of the Viewport Ui setup indicating the duplicate icon we're using at the moment and what kind of icon we'd like to represent the mode. That would be super helpful and then hopefully @amzn-leenguy can attach it to this discussion and we can add it to the build. Thanks!

**Reply by pollend** [reply]
here is how the current set of icons: 
![image](https://user-images.githubusercontent.com/854359/198696527-4fd07d57-d61b-446c-8c83-cbf523c068e5.png)

**Reply by hultonha** [reply]
Thanks @pollend, so we need a new icon for the third one down I believe that indicated 'advanced manipulation'. Would you be able to help us out @amzn-leenguy? Thanks!

<img width="194" alt="Screenshot 2022-10-31 at 10 54 43" src="https://user-images.githubusercontent.com/82228511/198992342-8cf14c63-9392-4604-bd5f-a13e6c7d04b6.png">

**Reply by amzn-leenguy** [reply]
I attempt to make some icons tomorrow (11/4) for advanced manipulation. Stay tuned!

**Reply by hultonha** [reply]
Awesome thanks @amzn-leenguy!

**Reply by amzn-leenguy** [reply]
Advanced Manipulation.zip
[Advanced Manipulation.zip](https://github.com/o3de/o3de/files/10026826/Advanced.Manipulation.zip)

**Reply by hultonha** [reply]
Fantastic thanks @amzn-leenguy! Unfortunately @pollend has had to take a brief hiatus from O3DE contributions but will hopefully be back soon 😝 In the meantime I might see if I have time to add this if @pollend doesn't have the opportunity. Thanks very much for the icon @amzn-leenguy! 🙂

**Reply by amzn-leenguy** [reply]
No problems. Its in SVG format. So we align with our standards. Sorry for the delay! I had a few things on my plate :)

**Comment by hultonha** [comment]
I've also mentioned @amzn-leenguy and @o3de/sig-ui-ux on this issue - https://github.com/o3de/o3de/issues/8269 (related to the visibility option on the White Box Component (tagging @pollend and @jonawals for visibility). Please see my comment https://github.com/o3de/o3de/issues/8269#issuecomment-1293786960 for further details. Thanks!

**Comment by mythrz** [comment]
e.g.,
**White Box VS Cylinder Shape**
It should show the same variables
https://www.o3de.org/docs/user-guide/components/reference/shape/cylinder-shape/

Also, the number of vertices in round circles such as the cylinder allows the creation of other shapes, like a hexagon.
Blender e.g., https://www.youtube.com/watch?v=aJGEqgUY1x8

Input controls through the keyboard, besides the mouse, could give extra control. Or Measurements.

**Reply by hultonha** [reply]
Hey @mythrz, I like the idea but White Box and Cylinder Shape are two very different types of Component meant for different things, having them show the same variables would be tricky given the many other shapes you could make with White Box. To change the number of vertices, have a look at the python scripts in the White Box Gem for shape generation, it may be possible to tweak them for your use case or write your own (e.g. There's an example staircase.py script)

**Reply by mythrz** [reply]
true. it's the sides.:
def create_cylinder(whiteBoxMesh, sides=16, size=0.5):

Can't find the UI thought/: Wondering if I can wire the variables myself. Any ideas?

**Reply by hultonha** [reply]
Hmm so I don't think this is exposed to the UI at the moment, but it might be possible to add this depending on the what shape type is selected in the Editor White Box Component drop down (see `DefaultShapeType` in `EditorWhiteBoxComponent`. It might be possible to display different UI depending on what's selected there (see the `EditContext` logic for how to do this if you're interested).

**Comment by pollend** [comment]
@amzn-leenguy where did we leave off on this, I had approval from my employer and should be available to work on this.

**Reply by AMZN-daimini** [reply]
@bhanuja-s @rainbj FYI :D

**Reply by bhanuja-s** [reply]
Heya @pollend, that's good to hear! I'll be taking over from @amzn-leenguy, and would love some context into your requirement. Thanks!

**Reply by pollend** [reply]
where is the plugin currently with the last changes I made :?. I believe I added manipulators, was that ever fully enabled or is it still in this preliminary state behind a CVAR I think? I think I was planning on looking at the plugin over the weekend. next week might be a bit hectic so I'll put in more time after next week.

**Reply by AMZN-daimini** [reply]
I had to port the Component Modes to use the new Action Manager system a few months ago, and in that process I tested and ported this as well. It's still in development and stabilization, but behind a settings registry property disabled by default.
